rfc/Errors-Extension-Specification.md

```diff
@@ -0,0 +1,123 @@
+
+```json
+{
+  "errors": [
+    {
+      "message": "Cannot query field 'foo' on type 'Bar'",
+      "path": [
+        "foo"
+      ],
+      "extensions": {
+        "code": "GRAPHQL_VALIDATION_FAILED"
+      }
+    }
+  ]
+}
+```
+
+With this information, the client can now handle the error more gracefully.
+
+However, this leads to the problem that Subgraph owners might send "anything" in the `extensions` field,
+which might leak sensitive information.
+
+Another problem is that users would like to standardize the error codes across different Subgraphs,
+e.g. `GRAPHQL_VALIDATION_FAILED` should always mean the same thing,
+and there should be agreement across all Subgraph owners on what codes to use.
+
+Additional fields in the `extensions` object could also be allowed, e.g. `message` or `severity`,
+but it's important that all Supergraph contributors agree on what fields and values are allowed.
+
+## Solution
+
+We propose to introduce the ability to define a schema for the `extensions` field in the Supergraph.
+As the response encoding is JSON, we can leverage JSON Schema to define the schema for the error object.
+
+The schema can be attached at the Supergraph level using the `wgc` cli tool using the following command:
+
+```bash
+npx wgc federated-graph create production -r http://router.example.com/graphql --label-matcher team=A department=backend --errors-extensions-schema ./errors-extensions-schema.json
+```
+
+Example of updating the schema for an existing Supergraph:
+
+```bash
+npx wgc federated-graph update production --errors-extensions-schema ./errors-extensions-schema.json
+```
+
+The schema file should look like this:
+
+```json
+{
+  "$schema": "http://json-schema.org/draft-07/schema#",
+  "type": "object",
+  "properties": {
+    "code": {
+      "type": "string",
+      "enum": [
+        "GRAPHQL_VALIDATION_FAILED",
+        "GRAPHQL_EXECUTION_FAILED"
+      ]
+    }
+  },
+  "required": [
+    "code"
+  ],
+  "additionalItems": false
+}
+```
+
+The `additionalItems` field is set to `false` to ensure that only the fields defined in the schema are allowed.
+
+In addition to just defining the schema,
+we also propose to add an additional option to define the behavior when the schema validation fails.
+
+When the JSON Schema validation fails for an `extensions` field,
+the Engine can either "skip" this field in the response object or "fail" the error rendering process and return a generic error message to the client.
+In addition, the Engine can log the error to the logs for debugging purposes and notify the Subgraph owner that their Subgraph is not returning errors in the correct format.
+
+The behavior can be defined in the Supergraph using the following command:
+
+```bash
+npx wgc federated-graph create production -r http://router.example.com/graphql --label-matcher team=A department=backend --errors-extensions-schema ./errors-extensions-schema.json --errors-extensions-validation-behavior skip
```

Comment:
Wouldn't it be more natural if the errors were pushed from the subgraphs? In that way, we could avoid coordination across teams. At composition time, we could merge the schemas. It also allows us to show what errors can be reported by which subgraph. Makes it easy to create connections with metrics.

Reply:
Isn't the idea to get one error schema for all errors? Sure, subgraphs could probably send in their respective codes and update that list on a per subgraph basis.

Reply:
I think merging schemas would be quite hard, for example the `required` field doesn't have any obvious semantics where it can be easily merged in a way that works for all subgraphs. 🤔

Reply:
I think the Schema needs to live on the Supergraph.
It's the only way to enforce coordination between teams and ensure that there's consensus on how errors should look like.